src/agent/rustjail/src/mount.rs

```diff
@@ -262,18 +281,54 @@ fn mount_cgroups(
     Ok(())
 }
 
-pub fn pivot_rootfs<P: ?Sized + NixPath>(path: &P) -> Result<()> {
+pub fn pivot_rootfs<P: ?Sized + NixPath + std::fmt::Debug>(path: &P) -> Result<()> {
     let oldroot = fcntl::open("/", OFlag::O_DIRECTORY | OFlag::O_RDONLY, Mode::empty())?;
     defer!(unistd::close(oldroot).unwrap());
     let newroot = fcntl::open(path, OFlag::O_DIRECTORY | OFlag::O_RDONLY, Mode::empty())?;
     defer!(unistd::close(newroot).unwrap());
-    unistd::pivot_root(path, path)?;
-    mount::umount2("/", MntFlags::MNT_DETACH)?;
     unistd::fchdir(newroot)?;
+    unistd::pivot_root(".", ".").chain_err(|| format!("failed to pivot_root on {:?}", path))?;
+    unistd::fchdir(oldroot)?;
+    mount::mount(
+        Some("none"),
+        ".",
+        Some(""),
+        MsFlags::MS_SLAVE | MsFlags::MS_REC,
+        Some(""),
+    )?;
+    mount::umount2(".", MntFlags::MNT_DETACH).chain_err(|| "failed to do umount2")?;
+    unistd::chdir("/")?;
     stat::umask(Mode::from_bits_truncate(0o022));
     Ok(())
 }
 
+fn rootfs_parent_mount_private(path: &str) -> Result<()> {
+    let mount_infos = parse_mount_table()?;
+
+    let mut max_len = 0;
+    let mut mount_point = String::from("");
+    let mut options = String::from("");
+    for i in mount_infos {
+        if path.starts_with(&i.mount_point) && i.mount_point.len() > max_len {
+            max_len = i.mount_point.len();
+            mount_point = i.mount_point;
+            options = i.optional;
```

Comment:
missing break?

Reply:
No,  here the loop try to find the mount point's parent mount,  by finding the longest mount point in the mount table.  So it would try to loop all of the entries here.